File: task_2/2(2).cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <set>
#include <cmath>
#include <random>

using namespace std;

// Перегрузка оператора вывода для вектора
template <typename T>
ostream& operator<<(ostream& os, vector<T>& result) {
    for (auto item : result) {
        os << "\t" << item << "\t║";  // Вывод каждого элемента с разделителями
    }
    os << "\n╠═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣\n";
    return os;
}

// Функция для возведения числа в степень по модулю
int powMod(int base, int degree, int mod) {
    int result = base % mod;  // Начальное значение
    for (int i = 2; i <= degree; i++) {
        result = (result * base) % mod;  // Умножаем и берем остаток от деления
    }
    return result;
}

// Решето Эратосфена для нахождения простых чисел до N
vector<int> sieveEratos(int N) {
    vector<bool> boolPrime(N + 1, true);  // Массив, в котором помечены все числа как простые
    for (int p = 2; p * p <= N; p++) {
        if (boolPrime[p]) {
            for (int i = p * p; i <= N; i += p) {
                boolPrime[i] = false;  // Отмечаем составные числа
            }
        }
    }

    vector<int> prime;
    for (int p = 2; p <= N; p++) {
        if (boolPrime[p]) {
            prime.push_back(p);  // Добавляем простые числа в результат
        }
    }

    return prime;
}

// Миллер-Рабин тест (проверка вероятности простоты числа)
bool rabin(int num, int k) {
    if (num == 2 || num == 3) {
        return true;  // 2 и 3 - простые
    }
    if (num < 2 || num % 2 == 0) {
        return false;  // Четные числа - составные
    }
    int s = 0;
    int d = num - 1;
    while (d % 2 == 0) {
        d /= 2;
        s++;  // Разлагаем (num - 1) на 2^s * d
    }
    for (int i = 0; i < k; i++) {
        int a = rand() % (num - 4) + 2;  // Случайное число для проверки
        int x = powMod(a, d, num);  // x = a^d mod num
        int y;
        for (int j = 0; j < s; j++) {
            y = powMod(x, 2, num);  // Вычисляем x^2 mod num
            if (y == 1 && x != 1 && x != (num - 1)) {
                return false;  // Если x = 1, то число составное
            }
            x = y;
        }
        if (y != 1) {
            return false;  // Если y != 1, то число составное
        }
    }
    return true;  // Число, скорее всего, простое
}

// Функция для генерации случайных чисел в заданном диапазоне
int randDist(int min, int max) {
    random_device rd;
    mt19937 rand(rd());
    uniform_int_distribution<int> dist(min, max);
    return dist(rand);
}

// Тест Поклингтона для проверки вероятности простоты числа
bool poklington(int n, int t, vector<int> qList) {
    set<int> aSet;
    int aj;

    // Генерируем t различных случайных чисел для теста
    while (aSet.size() != t) {
        aj = randDist(2, n - 1);
        aSet.insert(aj);
    }

    // Проверяем условие для всех a
    for (int aj : aSet) {
        if (powMod(aj, n - 1, n) != 1) {
            return false;
        }
    }

    // Проверяем дополнительные условия для делителей числа
    for (int aj : aSet) {
        bool allNol = true;
        for (int qi : qList) {
            if (powMod(aj, (n - 1) / qi, n) == 1) {
                allNol = false;
                break;
            }
        }
        if (allNol) {
            return true;  // n - вероятно простое
        }
    }

    return false;  // n - вероятно составное
}

// Функция для вычисления числа n и списка qList
tuple<int, vector<int>> calcN(vector<int> prime, int bit) {
    int maxIndex = prime.size() - 1, maxPow = 1;
    int min = bit / 2 + 1, max = bit / 2 + 2;
    int powMax = pow(2, max);
    for (; pow(2, maxPow) < pow(2, (bit / 2) + 1); maxPow++);

    int f = 1;
    vector<int> qList;

    // Генерация числа n и соответствующего списка qList
    while (true) {
        int num = randDist(0, maxIndex);  // Случайный индекс для простого числа
        int power = randDist(1, maxPow);  // Случайная степень для простого числа

        if (pow(prime[num], power)) {
            if (f * pow(prime[num], power) < powMax) {
                f *= pow(prime[num], power);  // Умножаем и обновляем f
                qList.push_back(prime[num]);  // Добавляем простое число в qList
            }
        }

        if (f > pow(2, (bit / 2))) {
            if (f >= pow(2, (bit / 2) + 1)) {
                f = 1;
                qList.clear();
            } else {
                break;
            }
        }
    }

    int R = randDist(min - 1, max - 1);  // Генерация случайного числа для R
    if (R % 2 != 0) {
        R++;
    }
    int n = R * f + 1;  // Вычисление числа n
    return make_tuple(n, qList);
}

int main() {
    int bit = 10, t = 10, n, k = 0;
    vector<int> prime = sieveEratos(500);  // Генерация простых чисел до 500

    vector<int> qList;
    vector<int> result;
    vector<string> testVer;
    vector<int> kCont;

    // Генерация чисел и их проверка
    while (result.size() < 10) {
        tie(n, qList) = calcN(prime, bit);  // Генерация числа n и списка qList
        if (poklington(n, t, qList)) {  // Проверка тестом Поклингтона
            result.push_back(n);

            if (rabin(n, 3)) {  // Проверка Миллера-Рабина
                testVer.push_back("+");
            } else {
                testVer.push_back("-");
            }
            kCont.push_back(k);
            k = 0;
        } else {
            if (rabin(n, 3)) {
                k++;
            }
        }
    }

    // Вывод результатов
    cout << "\n╔═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗\n";
    cout << "║Num\t║";
    for (int i = 1; i < 11; i++)
        cout << "\t" << i << "\t║";
    cout << endl;
    cout << "╠═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣";
    cout << endl;
    cout << "║P\t║" << result;
    cout << "║Test\t║" << testVer;
    cout << "║k\t║";
    for (auto item : kCont)
        cout << "\t" << item << "\t║";
    cout << endl;
    cout << "╚═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝";
}

File: task_2/2(2).go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// Функция для возведения числа в степень по модулю
func powMod(base, degree, mod int) int {
	result := base % mod
	// Возводим основание в степень по модулю с использованием метода быстрого возведения в степень
	for i := 2; i <= degree; i++ {
		result = (result * base) % mod
	}
	return result
}

// Решето Эратосфена для нахождения всех простых чисел до N
func sieveEratos(N int) []int {
	boolPrime := make([]bool, N+1)
	// Инициализируем все числа как простые
	for i := 2; i <= N; i++ {
		boolPrime[i] = true
	}

	// Основной цикл решета Эратосфена для исключения составных чисел
	for p := 2; p*p <= N; p++ {
		if boolPrime[p] {
			// Если число простое, помечаем все его кратные как составные
			for i := p * p; i <= N; i += p {
				boolPrime[i] = false
			}
		}
	}

	// Формируем список всех простых чисел
	primes := []int{}
	for p := 2; p <= N; p++ {
		if boolPrime[p] {
			primes = append(primes, p)
		}
	}

	return primes
}

// Генерация случайного числа в диапазоне [min, max]
func randDist(min, max int) int {
	rand.Seed(time.Now().UnixNano()) // Инициализация генератора случайных чисел
	return rand.Intn(max-min+1) + min
}

// Тест Рабина-Миллера на простоту
func rabin(num, k int) bool {
	if num == 2 || num == 3 {
		return true // 2 и 3 - простые
	}
	if num < 2 || num%2 == 0 {
		return false // Четные числа (кроме 2) составные
	}

	// Разлагаем число num-1 как 2^s * d
	s := 0
	d := num - 1
	for d%2 == 0 {
		d /= 2
		s++
	}

	// Повторяем k раз для случайных значений a
	for i := 0; i < k; i++ {
		a := rand.Intn(num-4) + 2
		x := powMod(a, d, num) // a^d % num
		y := 0
		for j := 0; j < s; j++ {
			y = powMod(x, 2, num) // Возведение в квадрат
			if y == 1 && x != 1 && x != (num-1) {
				return false // Если x = 1 или x = num-1, продолжаем, иначе число составное
			}
			x = y
		}
		if y != 1 {
			return false // Если не получилось получить 1, число составное
		}
	}
	return true // Число вероятно простое
}

// Тест Поклингтона на простоту
func poklington(n, t int, qList []int) bool {
	aSet := make(map[int]bool)

	// Генерируем t случайных чисел a, где 2 <= a < n
	for len(aSet) < t {
		a := randDist(2, n-1)
		aSet[a] = true
	}

	// Проверяем условие для каждого a: a^(n-1) % n == 1
	for a := range aSet {
		if powMod(a, n-1, n) != 1 {
			return false
		}
	}

	// Проверяем условие для каждого a, чтобы ни одно из a^(n-1/q) не было равно 1 для q из qList
	for a := range aSet {
		allNotOne := true
		for _, q := range qList {
			if powMod(a, (n-1)/q, n) == 1 {
				allNotOne = false
				break
			}
		}
		if allNotOne {
			return true // Если все условия выполнены, число вероятно простое
		}
	}

	return false // Если не выполнены условия для любого a, число составное
}

// Генерация числа n и списка простых множителей n-1
func calcN(primes []int, bit int) (int, []int) {
	maxIndex := len(primes) - 1
	maxPow := 1
	for ; math.Pow(2, float64(maxPow)) < math.Pow(2, float64(bit/2+1)); maxPow++ {
	}

	min := bit/2 + 1
	max := bit/2 + 2
	powMax := math.Pow(2, float64(max))

	f := 1
	qList := []int{}

	// Генерация n как произведения простых чисел
	for {
		randIndex := randDist(0, maxIndex)
		alphaDegree := randDist(1, maxPow)

		if math.Pow(float64(primes[randIndex]), float64(alphaDegree)) > 0 {
			if float64(f)*math.Pow(float64(primes[randIndex]), float64(alphaDegree)) < powMax {
				f *= int(math.Pow(float64(primes[randIndex]), float64(alphaDegree)))
				qList = append(qList, primes[randIndex])
			}
		}

		// Проверка на достижение нужного диапазона для числа
		if float64(f) > math.Pow(2, float64(bit/2)) {
			if float64(f) >= math.Pow(2, float64(bit/2+1)) {
				f = 1
				qList = []int{}
			} else {
				break
			}
		}
	}

	// Генерация случайного числа R
	R := randDist(min-1, max-1)
	if R%2 != 0 {
		R++ // Делаем R четным
	}
	n := R*f + 1 // Получаем число n

	// Удаляем дубликаты из списка простых множителей
	unique := make(map[int]bool)
	var uniqueQList []int
	for _, q := range qList {
		if !unique[q] {
			unique[q] = true
			uniqueQList = append(uniqueQList, q)
		}
	}

	return n, uniqueQList
}

func main() {
	bit := 10 // Битовая длина числа
	t := 10 // Количество тестов для Поклингтона
	primes := sieveEratos(500) // Получаем простые числа до 500

	var result []int // Список найденных простых чисел
	var testVer []string // Результаты теста Рабина-Миллера
	var kCont []int // Счётчик числа успешных тестов

	k := 0
	// Ищем 10 вероятно простых чисел
	for len(result) < 10 {
		n, qList := calcN(primes, bit) // Генерация числа n

		// Проверяем число с помощью теста Поклингтона
		if poklington(n, t, qList) {
			result = append(result, n)

			// Проверяем число с помощью теста Рабина-Миллера
			if rabin(n, 3) {
				testVer = append(testVer, "+")
			} else {
				testVer = append(testVer, "-")
			}
			kCont = append(kCont, k)
			k = 0
		} else {
			if rabin(n, 3) {
				k++ // Увеличиваем счётчик, если тест Рабина-Миллера прошёл
			}
		}
	}

	// Вывод результатов в виде таблицы
	fmt.Println("\n╔═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗")
	fmt.Print("║Num\t║")
	for i := 1; i < 11; i++ {
		fmt.Printf("\t%d\t║", i)
	}
	fmt.Println("\n╠═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣")

	fmt.Print("║P\t║")
	for _, item := range result {
		fmt.Printf("\t%d\t║", item)
	}
	fmt.Println("\n╠═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣")

	fmt.Print("║Test\t║")
	for _, item := range testVer {
		fmt.Printf("\t%s\t║", item)
	}
	fmt.Println("\n╠═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╣")

	fmt.Print("║k\t║")
	for _, item := range kCont {
		fmt.Printf("\t%d\t║", item)
	}
	fmt.Println("\n╚═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝")
}
